pyccel/ast/builtin_objects/list_functions.py

```diff
@@ -0,0 +1,80 @@
+# coding: utf-8
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+"""
+The List container has a number of built-in methods that are 
+always available.
+
+This module contains objects which describe these methods within Pyccel's AST.
+"""
+
+from pyccel.ast.datatypes import NativeVoid, NativeGeneric, NativeHomogeneousList
+from pyccel.ast.internals import PyccelInternalFunction, get_final_precision
+
+
+class ListAppend(PyccelInternalFunction):
+    """
+    Represents a call to the .append() method.
+
+    Represents a call to the .append() method of an object with a list type,
+    which adds an element to the end of the list. This method returns `None`.
+    The append method is called as follows:
+
+    >>> a = [1]
+    >>> a.append(2)
+    >>> print(a)
+    [1, 2]
+
+    Parameters
+    ----------
+    lst_bound_arg : TypedAstNode
+        The variable representing the list.
+    
+    new_elem : TypedAstNode
+        The argument passed to append() method.
+    """
+    __slots__ = ("_lst_bound_arg", "_append_arg")
+    _dtype = NativeVoid()
+    _shape = None
+    _order = None
+    _rank = 0
+    _precision = -1
+    _class_type = NativeHomogeneousList()
+    name = 'append'
+
+    def __init__(self, lst_bound_arg, new_elem) -> None:
+        list_precision = get_final_precision(lst_bound_arg)
+        check = (getattr(lst_bound_arg, 'order', None) == getattr(new_elem, 'order', None))
+        conditions = (
+            new_elem.dtype is not NativeGeneric() and
+            lst_bound_arg.dtype == new_elem.dtype and
+            list_precision == get_final_precision(new_elem) and
+            lst_bound_arg.rank - 1 == new_elem.rank and
+            (check if check else (lst_bound_arg.class_type == new_elem.class_type))
```

Comment:
Why do you only check if the class types match if the order is different?

Reply:
it was probably because of this case :
```python
a = [[1, 2, 3]]
a.append([1, 2])
a.append((2, 3))
```
in both append methods, the conditions for the homogeneity without the class_type checking would be true(both have `order: None`), so there was no other way to differentiate between the tuple as argument or the list as argument

Reply:
But isn't `check` true here? `None == None`. I'm not sure why the order and class type should be connected in any way

Reply:
The order of `a` a 2D list is 'C' ,the order of a 1D list/tuple is None, and they all have the same `dtype`, `rank`, precision, that's why i needed to check the class type,if not, appending a tuple to a list of lists would be possible

Reply:
My question isn't so much why you check the class type but why you **only** do this under certain circumstances? For example how do you think this code would behave if I try to append a cuda array to a list of numpy arrays?

Reply:
I don't really know the expected behavior with cuda arrays, but my understanding, going into implementing this method, is to try as much as possible to notice any pattern for homogeneity. In my case, it's related to the information that is accessible in my `ListAppend`: dtype, rank, precision, and order. However, sometimes, this information is not enough to check for homogeneity in the case of embedded containers. Although I realize this is a bit superficial, it's necessary to implement a mechanism to check the class types of the incoming container and compare them with the class types of the list.

Reply:
The future cuda arrays will be very similar to numpy arrays. The only property in `TypedAstNode` which will be different is `class_type`

Reply:
Ahhhh, ok I understand. Does that mean that your implementation doesn't handle this:
```python
a = [(1, 2), (3,4)]
a.append((2, 3))
```
?

Reply:
If it's going to be homogeneous in C/Fortran, then that check would no longer be needed, just unnecessary complexity

Reply:
I agree, that check would in fact prevent us from going the homogeneous in C/Fortran route. But I think we are better checking like Python. This matches what you have written in your tests and it will make life easier for @smazouz42 and @bauom as I suspect they will need to use `class_type` to differentiate between memory on CPU and memory on GPU

Reply:
I will have to check with them when they need such a thing; until then, the current implementation will be okay. We can always change the checks based on the future features

Reply:
I disagree. It is much easier to change this when designing the implementation than to go back through and fix it.

Also I still don't understand the logic of this code.

Also
> until then, the current implementation will be okay.

So you can handle this code:
```python
a = [(1, 2), (3,4)]
a.append((2, 3))
```
?

Reply:
With the deleted check, yes, I can. I also discovered that this logic doesn't allow for the following
```python
import numpy as np
a = [np.array([1, 2, 3], order='C'), np.array([1, 2, 4], order='C')]
a.append(np.array([1, 2,], order='C'))
```
As I have no access to the order, so that logic is not right

Reply:
Ok. It looks like you aren't checking the order anymore though. The following is inhomogeneous but not handled by your code:
```python
import numpy as np
a = [np.array([1, 2, 3], order='C'), np.array([1, 2, 4], order='C')]
a.append(np.array([1, 2,], order='F'))
```

Reply:
yes that's my issue that i described in the previous comment:
i don't have access to the order of the passed argument
![order](https://github.com/pyccel/pyccel/assets/66710845/9aa3ea2a-a6c4-4cb3-b466-d550a06811fd)

Reply:
this case as well:
```python
a = [[1, 2, 3]]
a.append([4, 5])
```
`a` has an `order` of `C`, and the passed argument has an `order` of `None`

Reply:
Ah crap, I think this may be why I ended up not implementing `np.ndarray[int]` as a datatype instead of `np.ndarray`. Do you think you have enough infrastructure to handle everything you need for now? If you can't work out the order of an element of a list then I think we need to disallow saving `np.ndarray` with rank>1 in lists.

My impression is that it is possible to get a working implementation with certain restrictions at the moment. But if we want the most flexible method possible then the datatypes need an overhaul. There should be 2 types of datatype:
- base types : these should hold enough information to describe the datatype in C, so
    - datatype
    - precision
- container types : these should hold enough information to describe the datatype in C, so
    - rank
    - order
    - base type

The `TypedAstNode` would then just save a container type. I can create an issue for this. I agree that it is out of scope for your PR

Reply:
Can't we just base the `order` on the targeted language? Meaning, even if the user specified the `order` as 'C' and wanted to convert to Fortran, the `order` would still be 'F'. Similarly, if someone wanted to convert to 'C' and specified the `order` as 'F', the order would still be 'C'.Or will there be a difference, memory wise, in C and Fortran if the user specified otherwise?

Reply:
> Can't we just base the order on the targeted language? 

No. The order is unrelated to the target language. See : https://github.com/pyccel/pyccel/blob/devel/developer_docs/order_docs.md

> Or will there be a difference, memory wise, in C and Fortran if the user specified otherwise?

Yes there will be a memory difference

Reply:
I think we should open an issue for this and leave it as it is for now.

Reply:
Nothing. I think raising the error is out of scope for this issue and rewriting the TypedAstNode definitely is. I will open 2 issues. One easy one that should be done sooner rather than later to raise an error and a much larger one to clean up the typing. This will be done later if we think it is useful

Reply:
See #1721 and #1722

Reply:
okey, so this PR should wait until the issue related to raising an error is fixed?

Reply:
No. It just shouldn't be put in a release until the error is raised. It can be marked as ready